numba/targets/arraymath.py

```diff
@@ -914,6 +914,141 @@ def nanmedian_impl(arry):
 #----------------------------------------------------------------------------
 # Element-wise computations
 
+@overload(np.tri)
+def np_tri(M, N=None, k=0):
+
+    # numpy behaviour is unexpected and unreasonable in the case where k is not
+    # an integer: for example, the following assertion fails:
+    #
+    #   assert np.tri(5, 6, k=-2.8).shape == (5, 6)
+    #
+    # rather than try to replicate, we explicitly require k to be an integer.
+    if not isinstance(k, types.Integer):
+        def _abort_mission(*args):
+            raise ValueError('k must be an integer')
+        return _abort_mission
+
+    @register_jitable
+    def _tri_impl(shape, k):
+        out = np.empty(shape, dtype=np.float64)  # numpy default dtype
+
+        for i in range(shape[0]):
+            m_max = min(max(0, i + k + 1), shape[1])
+            out[i, :m_max] = 1
+            out[i, m_max:] = 0
+
+        return out
+
+    def np_tri_impl(M, N=None, k=0):
+        return _tri_impl((M, N), k)
+
+    def np_tri_impl_no_n(M, N=None, k=0):
+        return _tri_impl((M, M), k)
+
+    if N in (None, types.none):
+        return np_tri_impl_no_n
+    else:
+        return np_tri_impl
+
+@register_jitable
+def _tessellate_mask(mask, shape):
```

Comment:
Yeah, I think it will work fine and is a better idea.

As I recall, Numpy will let you supply a 'condition' (m, n) which maps onto x (a, b,  .. m, n) and y (a, b, .. m, n) - but the jitted version of np.where does not and requires 'condition' to be dimensioned the same as x and y.

So you'd either need to manufacture a mask of the required dimensions, or mask the sub-array and build up the higher dimensions.

My initial approach at the latter didn't perform well, but I think I was just taking the wrong approach on reflection.  In my local tests, it was always slower than just making the higher dimension mask.

I've adopted the above in my last commit & will throw more tests at it.  It's another inspiring suggestion!  Brilliant stuff.